Engine/src/Eri/Utils/Events/VectorEvents/VectorEvents.cpp

```diff
@@ -0,0 +1,549 @@
+  _keyPressSubscribers.insert(index, sub);
+
+  _log->LogDebug("Subscriber %d added to KeyPress", _keyPressCount);
+  return _keyPressCount++;
+}
+
+u64 VectorEvents::SubscribeMouse(IMouseSubscriber *const sub)
+{
+  auto index = _mouseSubscribers.begin() + _mouseCount;
+  _mouseSubscribers.insert(index, sub);
+
+  _log->LogDebug("Subscriber %d added to Mouse", _mouseCount);
+  return _mouseCount++;
+}
+
+u64 VectorEvents::SubscribeWindowState(IWindowStateSubscriber *const sub)
+{
+  auto index = _windowStateSubscribers.begin() + _internalCount;
+  _windowStateSubscribers.insert(index, sub);
+
+  _log->LogDebug("Subscriber %d added to Internal", _internalCount);
+  return _internalCount++;
+}
+
+bool VectorEvents::UnsubscribeKeyPress(u64 id)
+{
+  auto index = _windowStateSubscribers.begin() + id;
+  if (index >= _windowStateSubscribers.end())
+  {
+      _log->LogDebug("Subscriber %d not found in Key Press", id);
+      return false;
+  }
+
+
+  _log->LogDebug("Subscriber %d removed from Key Press", id);
+  // this *is* costly I think... we'll make a map change sometime
+  _windowStateSubscribers.erase(index);
+  return true;
+}
+
+bool VectorEvents::UnsubscribeMouse(u64 id)
+{
+  auto index = _mouseSubscribers.begin() + id;
+  if (index >= _mouseSubscribers.end())
+  {
+    _log->LogDebug("Subscriber %d not found in Mouse", id);
+    return false;
+  }
+
+
+  _log->LogDebug("Subscriber %d removed from Mouse", id);
+  // this *is* costly I think... we'll make a map change sometime
+  _mouseSubscribers.erase(index);
+  return true;
+}
+
+bool VectorEvents::UnsubscribeWindowState(u64 id)
+{
+  auto index = _windowStateSubscribers.begin() + id;
+  if (index >= _windowStateSubscribers.end())
+  {
+    _log->LogDebug("Subscriber %d not found in Internal", id);
+    return false;
+  }
+
+
+  _log->LogDebug("Subscriber %d removed from Internal", id);
+  // this *is* costly I think... we'll make a map change sometime
+  _windowStateSubscribers.erase(index);
+  return true;
+}
+
+bool VectorEvents::PublishKeyPress(enum Keys key, enum KeyPressType keyPressType) const
+{
+  if (_keyPressSubscribers.empty())
+  {
+    return false;
+  }
+
+  for (auto subscribers : _keyPressSubscribers)
+  {
+    subscribers->onKeyEvent(keyPressType, key);
+  }
+
+  return true;
+}
+
+bool VectorEvents::PublishMouse(enum Mouse mouse, i32 x, i32 y) const
+{
+  if (_mouseSubscribers.empty())
+  {
+    return false;
+  }
+
+  for (auto subscribers : _mouseSubscribers)
+  {
+    subscribers->onMouseEvent(mouse, x, y);
+  }
+
+  return true;
+}
+
+bool VectorEvents::PublishWindowState(enum WindowState state, i32 left, i32 right) const
+{
+  if (_windowStateSubscribers.empty())
+  {
+    return false;
+  }
+
+  for (auto subscribers : _windowStateSubscribers)
+  {
+    subscribers->onWindowStateEvent(state, left, right);
+  }
+
+  return true;
+}
+
+void VectorEvents::RegisterLogger(ILogger *log)
+{
+  _log = log;
+}
+
+void VectorEvents::LogPress(enum Keys key) const
```

Comment:
Issue track this. I'm unsure what to do for this honestly. Where should this go???